### ISSUE TITLE
Refactor event based gateways

### ISSUE THREAD

**Issue by essweine** [sartography/SpiffWorkflow#419]
Event based gateways currently do not work properly with timers.

The implementation is a complete hack.  It would be better to restructure them the same way as was done for boundary events, so that the gateway actually occurs after the events in the task tree.  The library's internal workings are premised on tasks having orderly transitions (eg, parent completes, child becomes ready or waits; parents should *not* wait on their children) but that's not how event based gateways work and it makes them unreliable.

This should be pretty easy to fix, except for the difficulty of working with our parser.  So I'd like to put this off until https://github.com/sartography/SpiffWorkflow/issues/404 is complete (it's currently in progress).

**Comment by benzyc** [comment]
Agreed. THe event gateways do not work at all. It keeps waiting even the event of one branch is triggered. And it seems like do not work with message event at all

**Comment by essweine** [comment]
This PR https://github.com/sartography/SpiffWorkflow/pull/481 revisits event-based gateways, while addressing bugs that were apparently introduced with some other recent changes.  Now task state transitions and predictions are handled in line with the rest of the library, so I am closing this.
